diff --git a/dist/lib/src/compiler-path.js b/dist/lib/src/compiler-path.js
index ae33aa3028e1a120d9e84b043bb19a71f1083b96..3c8a3d442459ae51cf0a009f456c82b1d94126cb 100644
--- a/dist/lib/src/compiler-path.js
+++ b/dist/lib/src/compiler-path.js
@@ -24,6 +24,9 @@ function isLinuxMusl(path) {
 }
 /** The full command for the embedded compiler executable. */
 exports.compilerCommand = (() => {
+    // https://github.com/sass/embedded-host-node/issues/334#issuecomment-2366246771
+    const binPath = process.env.SASS_EMBEDDED_BIN_PATH;
+    if (binPath) return [binPath];
     const platform = process.platform === 'linux' && isLinuxMusl(process.execPath)
         ? 'linux-musl'
         : process.platform;
